tests/mcp_e2e_test.py

```
@@ -0,0 +1,577 @@
+    client.send_notification("notifications/initialized")
+    time.sleep(0.5)
+
+    # ─── Phase 1: tools/list ───
+    print("\n--- Phase 1: Tool Discovery ---")
+    resp = client.call("tools/list")
+    ok = resp is not None and "result" in resp
+    result("tools/list returns result", ok)
+    if ok:
+        tools = resp["result"].get("tools", [])
+        tool_names = sorted([t["name"] for t in tools])
+        result(f"tool count is 18", len(tools) == 18, f"got {len(tools)}: {tool_names}")
```

Ruff (0.15.9)

[error] f-string without any placeholders

Remove extraneous `f` prefix

(F541)